### ISSUE TITLE
react/no-unused-prop-types false positive on destructured props

### ISSUE THREAD

**Issue by victorias** [yannickcr/eslint-plugin-react#1002]
react/no-unused-prop-types thinks that both the onClick and children props are unused even though these are clearly being used:

~~~jsx
type Props = {
  onClick: () => void,
  children: any,
};

class Button extends Component {
  props: Props

  render () {
    const { props } = this;
    return (
      <button
        onClick={props.onClick}
      >
        {props.children}
      </button>
    );
  }
}

export default Button;
~~~

If I refer to onClick using this.props.onClick or destructure all the props from this.props it works.

**Comment by Augustin82** [comment]
We have been encountering the same bug.

**Comment by linuxonrails** [comment]
Same problem :-(

**Comment by DianaSuvorova** [comment]
flow as well. Need a label. @ljharb  I want to take a look into supporting flow soon. Gathering use cases. Thank you!

**Comment by DianaSuvorova** [comment]
first use case here is a known issue with the rule when using an intermediate variable to store props/parts of props.

Second is fixed on master.

**Comment by derwaldgeist** [comment]
Same if you use a prop by simple deconstructing like this:

```js
render({userId = 123}) {
  ...
}
```

**Comment by ljharb** [comment]
@derwaldgeist you can't use props like that; render takes no arguments.

**Comment by derwaldgeist** [comment]
Yep, just found this out myself 👍 I was turning a stateless component in a class and missed this.

**Comment by jseminck** [comment]
I've adding support for this in https://github.com/yannickcr/eslint-plugin-react/issues/1393 which is still WIP. Will add these test cases but I'm quite sure they work already.
